### PR TITLE
feat(unlock-js): add support for PublicLock v11

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This adds support for PublicLock V11  in unlock-js

Needs to be rebase once #9229 is merged

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: packages/unlock-js/src/PublicLock/v11/index.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import v10 from '../v10'
+import abis from '../../abis'
+
+export default {
+  version: 'v11',
+  PublicLock: abis.PublicLock.v11,
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/v11/index.js:6]: Wow, all functions identical? That's convenient :D

**Review by julien51 — COMMENTED** [review]
Do you think it would make sense to pass an expiration to the grantKeyExtension method to match grant keys?

**Comment by clemsos** [comment]
> Do you think it would make sense to pass an expiration to the grantKeyExtension method to match grant keys?

I quite like the simplicity of the function that way but that's just a code thing :) Not sure how passing a custom duration will impact things like renewal by cc.

**Comment by clemsos** [comment]
I thought the goal of `grantKeyExtension` was to mimic `extend` - without requiring a direct payment. The `extend` itself will extend by the amount of time specified in the lock. Internally they both use `_extendKey`, which just requires a tokenId. If we really need to pass the duration as a param to `grantKeyExtension` then it will require substantial changes to the contract under the hood. 

Also seems a good practice to have correspondance as  `grantKeyExtension` =  `extend` for key granters.

Then maybe there is a direct use case I dont know about that will require to extend a key by a given amount that differs from the lock duration.

**Comment by julien51** [comment]
Well, technically `grantKeyExtension` should be to `extend` what `grantKeys` is to `purchase`, so it should mimick _both_ `extend` and `grantKeys`.

I understand about the larger changes, so let's maybe hold on that and wait a bit.

**Review by julien51 — COMMENTED** [review]
Ca m'a l'air bien mais il y a un truc qui plante :/

**Comment by clemsos** [comment]
> Ca m'a l'air bien mais il y a un truc qui plante :/

Yes, keep hitting some sort of race conditions in tests. I fixed several but still happening for some reason

**Comment by clemsos** [comment]
> Any idea why this race condition only happens with v11?

We were reusing an existing account that already has a key (`accounts[11]` ) so we were reaching the max keys per address (added a condition to set more when v11). I tried similar tests directly in smart-contracts and it all works smoothly.

What is weird is that we weren't catching the custom error (we moved from `require` statements to custom errors in ) 

I bumped hardhat version, but it seems the errors arent exported in the abi so we have a 'VM Exception while processing transaction: reverted with an unrecognized custom error' instead of showing the actual error (should be `MAX_KEYS_REACHED`).

I added the custom errors to the lock ABI (following [ethers guidelines ](https://docs.ethers.io/v5/api/utils/abi/interface/#Interface--creating) but still not showing up :/

File: packages/unlock-js/src/abis.ts [hunk 2]
```diff
@@ -600,6 +600,39 @@ const abis = {
     "v11": {
       "contractName": "PublicLock",
       "abi": [
+        "error CANNOT_APPROVE_SELF()",
```
Comment (julien51) [inline packages/unlock-js/src/abis.ts:603]: That's cool!

**Review by julien51 — APPROVED** [review]

**Pull request merged**